### PR TITLE
[Kunlun]add xpu support for spawn

### PR DESCRIPTION
**Pull request by vslyu**
### PR types

New features 
### PR changes

Others
### Describe

add distributed.spawn support for multi xpu

build scripts:
```
#!/bin/bash

unset GREP_OPTIONS

branch=`git branch | grep \* | cut -d ' ' -f2`
place=xpu
build_type=release
rpc=grpc
testing=y
os=ubuntu
py_v=3.7
#py_v=2.7

while true ; do
  case "$1" in
    -b) branch="$2" ; shift 2 ;;
    -p) place="$2" ; shift 2 ;;
    -t) testing="$2" ; shift 2 ;;
    -build_type) build_type="$2" ; shift 2 ;;
    -rpc) rpc="$2" ; shift 2 ;;
    -os) os="$2" ; shift 2 ;;
    *)
       if [[ ${#1} > 0 ]]; then
          echo "not supported arugments ${1}" ; exit 1 ;
       else
           break
       fi
       ;;
  esac
done

if [ "${branch}" != "develop" ]; then
    branch="wxdev"
fi

echo "branch=${branch}"
echo "place=${place}"
echo "build_type=${build_type}"
echo "rpc=${rpc}"
echo "testing=${testing}"

#[[ $place == "gpu" || $place == "cpu" ]] || echo "not support ${place}" ; exit 1 ;
#[[ $testing == "yes" || $testing == "no" ]] || echo "not support ${testing}" ; exit 1 ;
#[[ $build_type == "release" || $build_type == "debug" ]] || echo "not support ${build_type}" ; exit 1 ;
#[[ $rpc == "brpc" || $rpc == "grpc" || $rpc == "brpcrdma" ]] || echo "not support ${rpc}" ; exit 1 ;

echo "last"

case "$place" in
    gpu) WITH_GPU=ON ;;
    cpu) WITH_GPU=OFF ;;
    xpu)
        WITH_XPU=ON
        WITH_GPU=OFF
        WITH_MKLML=OFF
        WITH_MKLDNN=OFF
        ;;
    *) echo "not support ${place}" ; exit 1 ;;
esac

case "$testing" in
    y) WITH_TESTING=ON ;;
    n) WITH_TESTING=ON ;;
    #n) WITH_TESTING=OFF ;;
    *) echo "not support ${testing}" ; exit 1 ;;
esac

case "$build_type" in
    release) CMAKE_BUILD_TYPE=Release ;;
    debug) CMAKE_BUILD_TYPE=RelWithDebInfo ;;
    *) echo "not support ${build_type}" ; exit 1 ;;
esac

case "$rpc" in
    brpc) WITH_GRPC=OFF ;;
    grpc) WITH_GRPC=ON ;;
    rdma) WITH_GRPC=OFF ; WITH_BRPC_RDMA=ON ;;
    *) echo "not support ${rpc}" ; exit 1 ;;
esac

case "$os" in
    cent)
        export LD_LIBRARY_PATH=/opt/_internal/cpython-2.7.11-ucs4/lib:${LD_LIBRARY_PATH#/opt/_internal/cpython-2.7.11-ucs2/lib:}
        export PATH=/opt/python/cp27-cp27mu/bin/:${PATH}
        PYTHON_FLAGS="-DPYTHON_EXECUTABLE:FILEPATH=/opt/python/cp27-cp27mu/bin/python"
        ;;
    ubuntu) ;;
    *) echo "not support ${os}" ; exit 1 ;;
esac

#build_dir=build_${os}_${branch}_${build_type}_${place}_${testing}_${rpc}
build_dir=build_${os}_${branch}_${build_type}_${place}_n_${rpc}_${py_v}
mkdir -p  ${build_dir}
cd ${build_dir}
third_party_dir=${os}_${build_type}_${place}

echo ${WITH_GPU}

set -x
#         -DCUDA_ARCH_NAME=Volta\
#         -DCUDA_ARCH_NAME=Pascal \
cmake ../../  -DTHIRD_PARTY_PATH=/home/liuyuhui/XPUPaddle/build/third_party/${third_party_dir}/ \
         -DWITH_MKLML=${WITH_MKLML:-ON} \
         -DWITH_MKLDNN=${WITH_MKLDNN:-ON} \
         -DWITH_GPU=${WITH_GPU:-ON} \
         -DWITH_XPU=${WITH_XPU:-OFF} \
         -DWITH_XPU_BKCL=${WITH_XPU:-OFF} \
         -DWITH_C_API=OFF \
         -DWITH_DISTRIBUTE=ON \
         -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE:-RelWithDebInfo} \
         -DWITH_TESTING=${WITH_TESTING:-ON} \
         -DWITH_STYLE_CHECK=OFF \
         -DWITH_CONTRIB=OFF \
         -DWITH_GRPC=${WITH_GRPC:-ON} \
         -DWITH_BRPC_RDMA=${WITH_BRPC_RDMA:-OFF} \
         -DWITH_FLUID_ONLY=ON \
         -DWITH_INFERENCE_API_TEST=OFF \
         -DCMAKE_INSTALL_PREFIX=/root/paddlebuild/${third_party_dir}/install \
         -DWITH_DOC=ON \
         -DCUDA_ARCH_NAME=Volta\
         -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
         -DWITH_PYTHON=ON \
         -DPY_VERSION=${py_v}
set +x

#if [[ $os == "cent" ]]; then
    #cd $build_dir
    make -j 20
#fi
```
cp build scripts into build sub dirs,  and execute it, requires attention about THIRD_PARTY_PATH, make it as your Paddle ROOT PATH.
example code:
```
from __future__ import print_function

import paddle
import paddle.nn as nn
import paddle.optimizer as opt
import paddle.distributed as dist

class LinearNet(nn.Layer):
    def __init__(self):
        super(LinearNet, self).__init__()
        self._linear1 = nn.Linear(10, 10)
        self._linear2 = nn.Linear(10, 1)

    def forward(self, x):
        return self._linear2(self._linear1(x))

def train(print_result=False):
    # 1. initialize parallel environment
    dist.init_parallel_env()

    # 2. create data parallel layer & optimizer
    layer = LinearNet()
    dp_layer = paddle.DataParallel(layer)

    loss_fn = nn.MSELoss()
    adam = opt.Adam(
        learning_rate=0.001, parameters=dp_layer.parameters())

    # 3. run layer
    inputs = paddle.randn([10, 10], 'float32')
    outputs = dp_layer(inputs)
    labels = paddle.randn([10, 1], 'float32')
    loss = loss_fn(outputs, labels)

    if print_result is True:
        print("loss:", loss.numpy())

    loss.backward()

    adam.step()
    adam.clear_grad()

# Usage 1: only pass function.
# If your training method no need any argument, and
# use all visible devices for parallel training.
if __name__ == '__main__':
    dist.spawn(train)

# Usage 2: pass function and arguments.
# If your training method need some arguments, and
# use all visible devices for parallel training.
if __name__ == '__main__':
    dist.spawn(train, args=(True,))

# Usage 3: pass function, arguments and nprocs.
# If your training method need some arguments, and
# only use part of visible devices for parallel training.
# If your machine hold 8 cards {0,1,2,3,4,5,6,7},
# this case will use cards {0,1}
if __name__ == '__main__':
    dist.spawn(train, args=(True,), nprocs=2)

# Usage 4: pass function, arguments, nprocs and gpus.
# If your training method need some arguments, and
# only use part of visible devices for parallel training, 
# you can pass `xpus` to select the XPU cards you want to use. 
# For example, this case will use cards {4,5}  if your machine hold more than 6 cards.
if __name__ == '__main__':
    dist.spawn(train, args=(True,), nprocs=2, xpus='4,5')
```
And results as following:
```
bkcl working thread func rank 0 bkcl dev id0
bkcl working thread func rank 1 bkcl dev id1
W0218 13:50:59.006999 18430 device_context.cc:187] Please NOTE: xpu device: 0
W0218 13:50:59.006999 18431 device_context.cc:187] Please NOTE: xpu device: 1
bkcl op start:1
bkcl op start:0
bkcl_broadcast6 argument: 1 0x2963f60 0x210100000 0x210100000 100 0 0 2
bkcl_broadcast6 argument: 0 0x23cbae0 0x210100000 0x210100000 100 0 0 2
bkcl op start:1
bkcl op start:0
bkcl_broadcast6 argument: 0 0x23cbae0 0x210120000 0x210120000 10 0 0 2
bkcl_broadcast6 argument: 1 0x2963f60 0x210120000 0x210120000 10 0 0 2
bkcl op start:1
bkcl op start:0
bkcl_broadcast6 argument: 1 0x2963f60 0x210140000 0x210140000 10 0 0 2
bkcl_broadcast6 argument: 0 0x23cbae0 0x210140000 0x210140000 10 0 0 2
bkcl op start:1
bkcl op start:0
bkcl_broadcast6 argument: 0 0x23cbae0 0x210160000 0x210160000 1 0 0 2
bkcl_broadcast6 argument: 1 0x2963f60 0x210160000 0x210160000 1 0 0 2
loss: [7.6320724]
loss: [2.853292]
bkcl op start:1
bkcl op start:0
```

### CODE REVIEW

File: python/paddle/distributed/spawn.py [hunk 1]
```diff
@@ -117,39 +123,75 @@ def _get_subprocess_env_list(nprocs, options):
...
+                    "correct `nprocs` and `gpus` arguments are passed." %
+                    (len(selected_gpu_list), nprocs))
+            for card_id in selected_gpu_list:
+                if card_id not in env_devices_list:
+                    raise ValueError("The selected gpu card %s cannot found in "
+                                     "CUDA_VISIBLE_DEVICES (%s)." %
+                                     (card_id, ",".join(env_devices_list)))
+
+    if core.is_compiled_with_xpu():
+        args.selected_gpus = options.get('xpus', None)
```
Comment (QingshuChen) [inline python/paddle/distributed/spawn.py:162]: args.selected_xpus

File: python/paddle/distributed/spawn.py [hunk 2]
```diff
@@ -117,39 +123,75 @@ def _get_subprocess_env_list(nprocs, options):
...
+        if args.selected_xpus is None:
+            args.selected_gpus = options.get('selected_gpus', None)
```
Comment (QingshuChen) [inline python/paddle/distributed/spawn.py:164]: 这个逻辑什么意思？为什么selected_xpus是空要设置selected_gpus?

File: python/paddle/distributed/spawn.py (hunk 1 above)
Comment (vslyu) [reply python/paddle/distributed/spawn.py:162]: 这里XPU启动的卡数用的还是args.selected_gpus和utils.py里面的逻辑一致，后续会改成args.selected_devices，稍后我改下

File: python/paddle/distributed/spawn.py (hunk 2 above)
Comment (vslyu) [reply python/paddle/distributed/spawn.py:164]: 写错了，这里应该是selected_gpus～

**Pull request closed without merging**